Update slack docs to reflect latest changes

Joerger opened the pull request:
- Update the slack docs to reflect https://github.com/gravitational/teleport-plugins/pull/433
- Reformat guide to be easier to follow
  - Moved install plugin to the top so that the identity directory already exists before we generate the identity files.
- Fix non-slack plugin roles by including `access_request.update`
- Update config example to reflect https://github.com/gravitational/teleport-plugins/blob/master/access/slack/example_config.toml
- Some other changes for clarity

Joerger committed:
Update slack docs to reflect latest changes.

ptgott commented:
@Joerger Just double-checking whether this is okay to review now or if I should wait for the `teleport-plugins` PR.

Joerger commented:
> @Joerger Just double-checking whether this is okay to review now or if I should wait for the `teleport-plugins` PR.

Up to you, if anything changes in the PR I can ping you for a re-review.

ptgott commented:
@Joerger Thanks! Let me know when conflicts in this branch are resolved and all the documentable behavior in [gravitational/teleport-plugins#433](https://github.com/gravitational/teleport-plugins/pull/433) is finalized, and I'll review this.

Joerger committed:
Update slack resources, other changes.

Joerger commented on examples/resources/plugins/teleport-slack.toml:
```diff
@@ -1,16 +1,40 @@
-# example slack plugin configuration TOML file
+# Copied from https://github.com/gravitational/teleport-plugins/blob/master/access/slack/example_config.toml
```
Would it be possible to reference https://github.com/gravitational/teleport-plugins/blob/master/access/slack/example_config.toml directly in our docs? Or will it only render if we have the file locally?

ptgott replied:
The remark plugin that evaluates partials replaces the `(!<path>!)` by [reading a local file](https://github.com/gravitational/docs/blob/1a972bbb3117d1845794b3a89a48cebd14973396/server/remark-includes.ts#L53), so you'll need to use a copy unfortunately.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
+### Install the Teleport Slack Plugin
 
-(!docs/pages/includes/plugins/rbac.mdx!)
+We recommend installing the Teleport Plugins alongside the Teleport Proxy. This is an ideal
```
I think we should move the following sentence into a `Notice`, which I'll include below: "We recommend installing the Teleport Plugins alongside the Teleport Proxy. This is an ideal location as plugins have a low memory footprint, and will require both public internet access and Teleport Auth access."

  ```markdown
  <Notice scope={["oss","enterprise"} type="tip">
  We recommend installing Teleport plugins on the same host as the Teleport
  Proxy Service. This is an ideal location as plugins have a low memory footprint
  and will require access to both the public internet and the Teleport Auth Service.
  </Notice>
  ```
  This will display this advice only for self-hosted users, since it doesn't apply to Teleport Cloud.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
+and Teleport Auth access. We currently only provide linux-amd64 binaries, you can also compile
+these plugins from [source](https://github.com/gravitational/teleport-plugins/tree/master/access/slack).
+
+**Install the plugin**
```
I don't think we need this line since we already have the H3 header above.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
+  To install from source you need `git` and `go >= (=teleport.golang=)` installed.
+
+  ```code
+  # Checkout teleport-plugins
````
```suggestion
  # Check out the teleport-plugins repository
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
+### Create a user and role for the plugin
 
-### Export the access-plugin certificate
+(!docs/pages/includes/plugins/slack-rbac.mdx!)
```
It looks like this include is only used here. Does this need to be an include?

ptgott commented on docs/pages/includes/plugins/identity-export.mdx:
```diff
@@ -1,4 +1,6 @@
-Teleport's plugins use the `access-plugin` role and user to approve access requests. We export the identity files to this plugin using [`tctl auth sign`](../../setup/reference/cli.mdx#tctl-auth-sign).
+Teleport's plugins use the `access-plugin` user to manage access requests. In order to act as
+the user, the plugin must present the Teleport Auth server with valid identity files.
```
```suggestion
the user, the plugin must present the Teleport Auth Service with valid identity files.
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
-### Export the access-plugin certificate
+(!docs/pages/includes/plugins/slack-rbac.mdx!)
+
+#### Export the access-plugin identity
```
```suggestion
### Export the access plugin identity
```

ptgott commented on docs/pages/includes/plugins/impersonations.mdx:
```diff
@@ -0,0 +1,26 @@
+  scopeOnly={true}
+  scope={["cloud"]}
+>
+Teleport Cloud requires authenticating with a role that has [`impersonation`](https://goteleport.com/docs/access-controls/guides/impersonation/) rights and can create the `access-plugin` role and user. Log in with `tsh` with a user that has this role or has a role with these `allow` rules.
```
Can we replace the "impersonation" link with a one-sentence explanation of what impersonation means here? Otherwise, a reader has to leave the page and possibly break focus to complete the guide.

After the code snippet, we can say something like, "Learn more about impersonation in [Impersonating Teleport Users](path)." 

Also, we should use the "../relative/path/to/link.mdx" format so we have more control over our docs links.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
 (!docs/pages/includes/plugins/identity-export.mdx!)
 
-We'll reference these files later when [configuring the plugins](#configuring-the-teleport-slack-plugin).
+The above sequence should generate PEM encoded file(s) with the `access-plugin` user's certificates.
```
This paragraph repeats information included in identity-export.mdx.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
+```
 
-We'll create a new Slack app and set up auth tokens and callback URLs, so that Slack knows how to notify the Teleport plugin when Approve / Deny buttons are clicked.
+### Create Slack App
````
```suggestion
### Create a Slack app
```
We use sentence case for headers.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -34,27 +34,70 @@ This guide assumes that you have:
 3. Obtain an OAuth token.
 
-#### Creating a new Slack app
+### Create a Slack app
```
Repeats the header above. Let's remove this.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
-location as plugins have a low memory footprint, and will require both public internet access
-and Teleport Auth access.  We currently only provide linux-amd64 binaries, you can also
-compile these plugins from [source](https://github.com/gravitational/teleport-plugins/tree/master/access/slack).
+Teleport Slack uses a config file in TOML format. Generate a boilerplate config by
```
```suggestion
The Teleport Slack plugin uses a config file in TOML format. Generate a boilerplate config by
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
-  ```
-</TabItem>
-</Tabs>
+(!examples/resources/plugins/teleport-slack-configure.toml)
````
Broken include

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
-</Tabs>
+(!examples/resources/plugins/teleport-slack-configure.toml)
 
+#### Editing the config file
```
```suggestion
### Editing the config file
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
+#### Editing the config file
 
-Run `./install` from `teleport-slack` or place the executable in the appropriate `/usr/bin` or `/usr/local/bin` on the server installation.
+We need to edit the config files with values specific to your Teleport and Slack App intances.
```
```suggestion
Edit the following fields in the configuration file you generated with values specific to your Teleport and Slack app instances.
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
+We need to edit the config files with values specific to your Teleport and Slack App intances.
 
-### Configuring the Teleport Slack plugin
+Open the config file we just generated - `/etc/teleport-slack.toml` - and you should see the following fields which we need to edit.
```
Don't need this sentence

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
 
-Teleport Slack uses a config file in TOML format. Generate a boilerplate config by
-running the following command:
+**`[teleport]`**
```
I think we should put all of the text under the `[teleport]` config option under a `Tabs` component so self-hosted/Cloud users don't see scope-irrelevant information. Currently, readers need to determine which information applies to them, and self-hosted users have to skip over the "Cloud Proxy Address" Admonition. Using one `Tabs` box for everything in the `[teleport]` discussion will make this simpler.

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -81,74 +124,83 @@ After adding to the workspace, you still need to invite the bot to the channel.
+
+**`[role_to_recipients]`**
+
+provide a mapping of Teleport roles to Slack recipients based on desired functionality.
```
```suggestion
Provide a mapping of Teleport roles to Slack recipients based on desired functionality.
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -164,40 +216,51 @@ INFO   Plugin is ready slack/app.go:101
+<TabItem label="As admin">
+  A Teleport admin can create an access request for another user with `tctl`.
+  ```code
+  # Replace USERNAME with a Teleport local user, and TARGET_ROLE with a Teleport Role
````
```suggestion
  # Replace USERNAME with a Teleport local user and TARGET_ROLE with a Teleport Role
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -164,40 +216,51 @@ INFO   Plugin is ready slack/app.go:101
+  ```
+</TabItem>
+<TabItem label="As user">
+  Users can use `tsh` to create an access request and login with approved roles.
````
```suggestion
  Users can use `tsh` to create an access request and log in with approved roles.
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -164,40 +216,51 @@ INFO   Plugin is ready slack/app.go:101
+<TabItem label="As user">
+  Users can use `tsh` to create an access request and login with approved roles.
+  ```code
+  # Replace TARGET_ROLE with a Teleport Role
````
```suggestion
  # Replace TARGET_ROLE with a Teleport role
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -164,40 +216,51 @@ INFO   Plugin is ready slack/app.go:101
+</Tabs>
 
-### Log in and request a role
+The configured slack recipients should receive a new message notification of a new pending request.
```
```suggestion
The configured Slack recipients should receive a new message notification of a new pending request.
```

ptgott commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -164,40 +216,51 @@ INFO   Plugin is ready slack/app.go:101
+</Tabs>
 
-You should now see a new request in Teleport, and a message about the request on Slack with instructions.
+Once the request is resolved, the slack bot will make an emoji reaction of ✅ or ❌ on the slack 
```
```suggestion
Once the request is resolved, the Slack bot will make an emoji reaction of ✅ or ❌ on the Slack 
```

ptgott replied:
Is "the Slack message" the access request message?

Joerger committed:
Make requested changes.

ptgott commented on examples/resources/plugins/teleport-slack.toml:
```diff
@@ -1,16 +1,40 @@
-# example slack plugin configuration TOML file
+# Copied from https://github.com/gravitational/teleport-plugins/blob/master/access/slack/example_config.toml
+
+# Example slack plugin configuration TOML file
```
```suggestion
# Example Slack plugin configuration TOML file
```

ptgott reviewed (commented):
I think this guide should be structured similarly to our other step-by-step guides. Each H2 section after the Prerequisites should be titled `Step <number>/<total steps>. <Perform this step>`. As much as possible, this guide should be a linear sequence of steps. This will make it clearer how much progress the user has made at each point in the guide.

I'm happy to create a PR against this branch (or push changes to it directly) to implement this structure if that makes sense for you!

Joerger commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
````diff
@@ -34,27 +34,69 @@ This guide assumes that you have:
+  $ tar -xzf teleport-access-slack-v(=teleport.version=)-linux-amd64-bin.tar.gz
+  ```
+  Run `./teleport-access-slack/install`, or manually place `./teleport-access-slack/teleport-slack`
+  into an appropriate location within the sytem's PATH. e.g. `/usr/bin`, `/usr/local/bin`.
````
Good point, I'll just add it to the code block and remove the extra sentence.

Joerger commented:
> I'm happy to create a PR against this branch (or push changes to it directly) to implement this structure if that makes sense for you!

@ptgott That'd be great, feel free to push directly to this branch, thanks!

Joerger committed:
Make requested changes.

ptgott commented:
@Joerger Sorry for the wait on this. I've carved out time later this week to push changes and get this merged.

marcoandredinis commented:
> @Joerger Sorry for the wait on this. I've carved out time later this week to push changes and get this merged.

I'm creating the same page for the email plugin
I'll wait for your PR before proceeding, otherwise we would have duplicated efforts in the review
(for reference, this is the PR https://github.com/gravitational/teleport/pull/11796)

ptgott committed:
Edit the Slack guide

- Extract commercial-only tabbed prerequisites into a partial
- Add a warning for OSS users
- Adopt the structure of other step-by-step guides, with H2s for
  Prerequisites and numbered steps
- Fix a link in the impersonation.mdx partial
- Misc copy-edits

ptgott commented:
@Joerger Just pushed edits!

ptgott commented:
@Joerger Another thing I forgot to mention: the "Audit log" section seems to be missing a word or two:

> The plugin will let anyone with access to the Slack Channel so it's
important to review Teleport' audit log.

What will the plugin let anyone with access to the Slack channel do? Why is it important to review the audit log?

Joerger commented:
> > The plugin will let anyone with access to the Slack Channel so it's
> > important to review Teleport' audit log.
> 
> What will the plugin let anyone with access to the Slack channel do? Why is it important to review the audit log?

I'm not sure, but it looks like that line is in every workflow guide.

marcoandredinis commented:
> @Joerger Another thing I forgot to mention: the "Audit log" section seems to be missing a word or two:
> 
> > The plugin will let anyone with access to the Slack Channel so it's
> > important to review Teleport' audit log.
> 
> What will the plugin let anyone with access to the Slack channel do? Why is it important to review the audit log?

I think that section is valid for Slack like integrations
It's kinda of a heads up that whoever has access to a particular channel will be able to approve the request
You just need to press the Approve button, there's no other validation (as far as I know)

However, for integrations like email, you'll only receive the link to the Teleport UI, and the webapp will validate whether or not you can approve the request (based on your role permissions).

So, having access to a Slack channel is all it takes to approve a request
Having access to the email is more like a FYI, a Notification, you still go trough the Teleport Roles validation before you can make a decision

ptgott committed:
Clarify why you would use the audit log

ptgott committed:
Respond to PR feedback

Joerger committed:
Fix broken link.

marcoandredinis commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -22,39 +32,74 @@ This guide will talk through how to set up Teleport with Slack. Teleport's Slack
+We currently only provide `linux-amd64` binaries. You can also compile these plugins
+from source.
+
+<Notice scope={["oss","enterprise"]} type="tip">
```
The first notice of this guide mentions we only support enterprise version
But here, we have `oss` as part of the scope
Should we remove `oss` here?

ptgott replied:
That was a mistake—removed.

marcoandredinis commented on docs/pages/enterprise/workflow/ssh-approval-slack.mdx:
```diff
@@ -4,9 +4,19 @@ description: This guide explains how to setup a Slack plugin for Teleport for pr
+  type="danger"
+  scope="oss"
+>
+  This guide requires a commercial edition of Teleport. The open source
```
Is commercial and enterprise the same thing?

ptgott replied:
In some docs guides we have used "commercial" to refer to "Cloud or Enterprise." I've made this a bit more specific here.

marcoandredinis commented on examples/resources/plugins/teleport-slack.toml:
```diff
@@ -1,16 +1,40 @@
-# Optional Slack Rooms
-recipients = ["team","devops"]
+# Slack Bot OAuth token
+token = "xoxb-11xx"
```
We should add that we support reading the token from a file
Related: https://github.com/gravitational/teleport-plugins/issues/505

marcoandredinis replied:
Not sure if we should add it here, in the doc itself or both

ptgott replied:
I've added this in a comment in the example file

ptgott committed:
Respond to PR feedback

marcoandredinis reviewed (approved):
LGTM
